src/adapter/src/coord/command_handler.rs

```diff
@@ -683,6 +688,87 @@ impl Coordinator {
+                        TimelineContext::TimelineDependent(timeline) => timeline,
+                        TimelineContext::TimestampDependent
+                        | TimelineContext::TimestampIndependent => {
+                            // We default to EpochMilliseconds, similarly to `determine_timestamp_for`.
+                            // Note that we didn't accurately decide whether we are TimestampDependent
+                            // or TimestampIndependent, because for this we'd need to also check whether
+                            // `query.contains_temporal()`, similarly to how `peek_stage_validate` does.
+                            // However, this doesn't matter here, as we are just going to default to
+                            // EpochMilliseconds in both cases.
+                            Timeline::EpochMilliseconds
+                        }
+                    };
+                    Some(self.get_timestamp_oracle(&timeline).read_ts().await)
+                    // TODO: It might be good to take into account `least_valid_read` in addition to
+                    // the oracle's `read_ts`, but there are two problems:
+                    // 1. At this point, we don't know which indexes would be used. We could do an
+                    // overestimation here by grabbing the ids of all indexes that are on ids
+                    // involved in the query. (We'd have to recursively follow view definitions,
+                    // similarly to `validate_timeline_context`.)
+                    // 2. For a peak, when the `least_valid_read` is later than the oracle's
+                    // `read_ts`, then the peak doesn't return before it completes at the chosen
+                    // timestamp. However, for a CRATE MATERIALIZED VIEW statement, it's not clear
+                    // whether we want to make it block until the chosen time. If it doesn't block,
+                    // then the initial refresh wouldn't be linearized with the CREATE MATERIALIZED
+                    // VIEW statement.
+                    //
+                    // Note: The Adapter is usually keeping a read hold of all objects at the oracle
+                    // read timestamp, so `least_valid_read` usually won't actually be later than
+                    // the oracle's `read_ts`. (see `Coordinator::advance_timelines`)
+                    //
+                    // Note 2: If we choose a timestamp here that is earlier than
+                    // `least_valid_read`, that is somewhat bad, but not catastrophic: The only
+                    // bad thing that happens is that we won't perform that refresh that was
+                    // specified to be at `mz_now()` (which is usually the initial refresh)
+                    // (similarly to how we don't perform refreshes that were specified to be in the
+                    // past).
+                } else {
+                    None
+                };
```

Comment:
In the meantime, the
```
                    let timeline = match timeline_context {
                        TimelineContext::TimelineDependent(timeline) => timeline,
                        TimelineContext::TimestampDependent
                        | TimelineContext::TimestampIndependent => {
                            // We default to EpochMilliseconds, similarly to `determine_timestamp_for`.
                            // Note that we didn't accurately decide whether we are TimestampDependent
                            // or TimestampIndependent, because for this we'd need to also check whether
                            // `query.contains_temporal()`, similarly to how `peek_stage_validate` does.
                            // However, this doesn't matter here, as we are just going to default to
                            // EpochMilliseconds in both cases.
                            Timeline::EpochMilliseconds
                        }
                    };
```
part got simplified into
```
                    // We default to EpochMilliseconds, similarly to `determine_timestamp_for`,
                    // but even in the TimestampIndependent case.
                    // Note that we didn't accurately decide whether we are TimestampDependent
                    // or TimestampIndependent, because for this we'd need to also check whether
                    // `query.contains_temporal()`, similarly to how `peek_stage_validate` does.
                    // However, this doesn't matter here, as we are just going to default to
                    // EpochMilliseconds in both cases.
                    let timeline = timeline_context
                        .timeline()
                        .unwrap_or(&Timeline::EpochMilliseconds);
```

If we were to allow this code to not support anything else then `EpochMilliseconds`, then we could further simplify by deleting even the
```
                    let timeline_context =
                        match self.validate_timeline_context(resolved_ids.0.clone()) {
                            Ok(tc) => tc,
                            Err(e) => return ctx.retire(Err(e)),
                        };
```
part, and then simply using `Timeline::EpochMilliseconds` in the `get_timestamp_oracle` call. But since this is already written, maybe it could just stay as it is, and then it would be a bit more future proof than simply assuming `EpochMilliseconds`?

Reply:
Ack, works for me.